aesd-char-driver/main.c

```diff
@@ -0,0 +1,190 @@
+/**
+ * @file aesdchar.c
+ * @brief Functions and data related to the AESD char driver implementation
+ *
+ * Based on the implementation of the "scull" device driver, found in
+ * Linux Device Drivers example code.
+ *
+ * @author Dan Walkes
+ * @date 2019-10-22
+ * @copyright Copyright (c) 2019
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/syscalls.h>
+#include <linux/init.h>
+#include <linux/printk.h>
+#include <linux/types.h>
+#include <linux/cdev.h>
+#include <linux/fs.h> // file_operations
+#include <asm/uaccess.h>
+#include "aesdchar.h"
+int aesd_major =   0; // use dynamic major
+int aesd_minor =   0;
+
+MODULE_AUTHOR("Swapnil Ghonge");
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+   
+    PDEBUG("open");
+    try_module_get(THIS_MODULE);
+
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    
+    PDEBUG("release");
+    
+    module_put(THIS_MODULE);
+
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    size_t  entry_offset = 0;
+    
+    
+    struct aesd_buffer_entry *entry = NULL;
+    PDEBUG("read %zu bytes with offset %lld\n",count,*f_pos);
+    
+    mutex_lock(&aesd_device.lock);
+    entry = aesd_circular_buffer_find_entry_offset_for_fpos(&aesd_device.buffer, *f_pos, &entry_offset);
```

Comment:
@swapnilghonge I think you are missing some logic here to actually read the circular buffer content.  I don't think this will work with drivertest or fulltest scripts.  Looks like you don't have a working implementation in https://github.com/cu-ecen-aeld/assignment-8-swapnilghonge.